ENH: Caching, advanced mapping and separating events for MISP Feed output bot

kamil-certat opened the pull request:
MISP Feed output bot got new features:

 * caching incoming messages to not re-generate on every run;
 * an ability to select which fields should be mapped and set additional parameters for attributes;
 * an ability to group messages in different MISP events, based on a field.

The bot is fully backward-compatible. By default, the previous behaviour is kept.

In addition, code related to Python 3.6 was removed and the message library was fixed not to modify the original dict instance.

This PR replaces on PR #2505.

kamil-certat committed:
ENH: Add possibility to delay generating MISP Feed

Generating MISP feed on every incoming message slows down processing.
The new config option let us decide to save them in batches. Cached
events are stored in a cache list in Redis. In addition, a code
related to Python 3.6 was removed as we do not support this version
any more.

kamil-certat committed:
Add documentation. Fix code compatibility

kamil-certat committed:
Fix spelling

kamil-certat committed:
ENH: Add attribute mapping

The bot can now construct an event much more alligned to custom
needs, allowing setting comments and selecting just a subset of
fields to export

kamil-certat committed:
ENH: Add support for creating separated MISP Events

With `event_separator` parameter, user can decide to create
more than one MISP event in the output bot and group incomming
messages based on given field.

kamil-certat commented:
Yeah, finally green :tada:

sebix commented:
Are you using it in production?

kamil-certat committed:
FIX: Handle not existing fields with manual mapping

kamil-certat commented:
We started using it on staging, found some pain points, and now I'll test it & hopefully promote to prod in a few days - so, not yet, but soon ;)

sebix commented:
There's not much I can check here without reading the MISP code and docs and setting up an instance.

Maybe @Rafiot, can you have a glimpse?

aaronkaplan commented on docs/user/bots.md:
```diff
@@ -4595,6 +4595,44 @@ The PyMISP library >= 2.4.119.1 is required, see
 () The output bot creates one event per each interval, all data in this time frame is part of this event. Default "1
 hour", string.
```
what happens when set to 0 ?

kamil-certat replied:
I've looked in the code, and it looks like in such case, this would be a ValueError. However, you can use "0 hours" what should result in creating a new MISP event for every IntelMQ event.

aaronkaplan commented on docs/user/bots.md:
```diff
@@ -4595,6 +4595,44 @@ The PyMISP library >= 2.4.119.1 is required, see
+would create a MISP object with three attributes `source.ip`, `feed.name` and `destination.ip`
+and set their values as in the IntelMQ event. In addition, the `feed.name` would have a comment
+as given in the `event_description.text` from IntelMQ event, and `destination.ip` would be set
+as not usable for IDS.
```
clear and understandable. Cool

aaronkaplan commented on docs/user/bots.md:
```diff
@@ -4595,6 +4595,44 @@ The PyMISP library >= 2.4.119.1 is required, see
+
+(optional, string): If set to a field name from IntelMQ event, the bot will group incoming messages
+in separated MISP events, based on the value of this field. The `interval_event` parameter acts
+for all grouping events together.
```
this is not 100% clear for me yet:

will it collect events until (... until when?...) and then sort and group by this viel?
I think rephrasing this slightly makes it clearer in the documentaiton.

kamil-certat replied:
What do you think about the current version?

> will it collect events until (... until when?...) and then sort and group by this viel?

This field is responsible only for separating using the field. Collecting & holding is a different setting, that can be used together, but you don't have to. Let's have a look at the new description

aaronkaplan commented on intelmq/bots/outputs/misp/output_feed.py:
```diff
@@ -5,37 +5,40 @@
-    output_dir: str = "/opt/intelmq/var/lib/bots/mispfeed-output"  # TODO: should be path
+    output_dir: str = (
+        "/opt/intelmq/var/lib/bots/mispfeed-output"  # TODO: should be path
+    )
```
why the tuple?

kamil-certat replied:
It's not a tuple ;) It's a Python way to put the string into next line (or split into multiple strings) to avoid line-length limit - although it's my fault. The rules we enforce in the repository are very weak, so weak that my almost every time wants tot reformat a lot, and sometimes I forget reverting changes not related to my code.

BTW, I think more and more that we would profit from enforcing black/ruff instead of weak pycodestyle.

aaronkaplan replied:
I am also using black more and more...

aaronkaplan commented on intelmq/bots/outputs/misp/output_feed.py:
```diff
@@ -65,80 +64,201 @@ def init(self):
+        if key in self.current_events:
+            event = self.current_events[key]
+        else:
+            event = self._generate_new_event(key)
```
don't we want to support at add a default tag to an event, such as "FEED:Source=intelmq" ?
Of course, this should be configured by the user.

kamil-certat replied:
It's an interesting request, could you please elaborate how do you see it? I was focused on what we need now (so I didn't include a few different things I was thinking about), but if you have a specific idea you think it would be important to include now - let's share.

aaronkaplan commented on intelmq/bots/outputs/misp/output_feed.py:
```diff
@@ -65,80 +64,201 @@ def init(self):
+            self.current_events = {}
 
-    def process(self):
+    def _load_event(self, file_path: Path, key: str):
```
missing function docstring ?

kamil-certat replied:
Not at all. If you wish to document `_load_event` with `This function loads an event from file`, feel free, but I disagree with enforcing a docstring requirement for every function. It should happen when the function is complex, otherwise it's useless. Over-commenting is not a good practice. If there is anything you don't understand in the function, please let me know.

aaronkaplan commented on intelmq/bots/outputs/misp/output_feed.py:
```diff
@@ -65,80 +64,201 @@ def init(self):
+
+        self.acknowledge_message()
+
+    def _generate_new_event(self, key):
```
missing function docstring?

kamil-certat replied:
Disagree it's needed here.

aaronkaplan commented on intelmq/bots/outputs/misp/output_feed.py:
```diff
@@ -65,80 +64,201 @@ def init(self):
+                json.dump({k: str(v) for k, v in self.current_files.items()}, f)
+        return self.current_events[key]
+
+    def _add_message_to_feed(self, message: dict):
```
missing func docstring?

kamil-certat replied:
Disagree it's needed here.

aaronkaplan commented on intelmq/bots/outputs/misp/output_feed.py:
```diff
@@ -65,80 +64,201 @@ def init(self):
+        else:
+            self._custom_mapping(obj, message)
+
+    def _default_mapping(self, obj: "MISPObject", message: dict):
```
docstring?

kamil-certat replied:
Disagree it's needed here.

aaronkaplan commented on intelmq/bots/outputs/misp/output_feed.py:
```diff
@@ -65,80 +64,201 @@ def init(self):
+                        object_relation,
+                    )
+
+    def _extract_misp_attribute_kwargs(self, message: dict, definition: dict) -> dict:
```
docstring? not just # comment

aaronkaplan commented on intelmq/bots/outputs/misp/output_feed.py:
```diff
@@ -65,80 +64,201 @@ def init(self):
+        return result
 
-        feed_output = self.current_event.to_feed(with_meta=False)
+    def _custom_mapping(self, obj: "MISPObject", message: dict):
```
same

kamil-certat replied:
Disagree it's needed here.

aaronkaplan commented on intelmq/bots/outputs/misp/output_feed.py:
```diff
@@ -65,80 +64,201 @@ def init(self):
 
-        with self.current_file.open('w') as f:
-            json.dump(feed_output, f)
+    def _generate_feed(self, message: dict = None):
```
same

kamil-certat replied:
Disagree it's needed here.

aaronkaplan commented on intelmq/lib/mixins/cache.py:
```diff
@@ -1,17 +1,29 @@
 
 
 class CacheMixin:
+    """Provides caching possibilities for bots
```
quick line (link to documentation on the mixins) ?

aaronkaplan reviewed (dismissed):
minor changes requested pls.

aaronkaplan commented:
My main request is:

1) give the possibility (or document it) how to add default tags to events (such as "FEED:Source=intelmq") to any event which gets generated via this feed. Adding a default tag helps in sorting out automatically added events and attributes in case something goes wrong.
2) make sure the users understand in the documentation how feeds work. Linking to the MISP documentation and what each parameter when including the MISP feed actually means.
For example, did you know that if  you set "caching:enabled" in a MISP feed, then that means the feed events and attributes don't get added to the `events` (and `attributes`) table in MISP. 
It's totally non-intuitive. 
Why am I saying this? I had to figure it out the hard way myself.
3) basically we need a *tutorial* on docs.intelmq.org  on how to run this properly!
It's a very cool feature and addition but you can quickly shoot yourself in the foot if you don't know how it works and what every parameter in MISP means.

kamil-certat commented:
Thanks for the review!

re 1: It's a good idea, I'll look into it. I was concentrated on what we actually need, and I do think there are more things we can improve in the bot.
re 2: It's a good point, I will point to the MISP documentation. I also had to explore what we really can do, the PyMISP documentation wasn't really helpful :D 
re 3: yeah, I agree it needs a tutorial (and I've planned it), but it's a separated thing, I won't include it here. My plan is to write a blog post once we establish the final integration internally, and then we could include it in the documentation as well. I expect to publish it around September, hopefully. There is a queue of what I'd like to write about, and a limited time ;)

kamil-certat committed:
ENH: Add option to extend default info

kamil-certat committed:
Fix typos

kamil-certat committed:
ENH: add support for tagging

kamil-certat committed:
Fix generating on restart

kamil-certat committed:
ENH: Add tagging, check, and improved docs

kamil-certat committed:
Adjust to pycodestyle

kamil-certat committed:
DOC: Update documentation about CacheMixin

kamil-certat committed:
Fix typo

kamil-certat committed:
Clean up imports in tests

kamil-certat commented:
@aaronkaplan 

Could you have a look again? I have implemented tagging as well as rewritten the documentation, added config examples, implemented validation in the `check` method and refactored code, including renaming methods to be more clear what they do or adding doc strings where I found it useful. I have also added descriptions to complex test cases explaining what should happen.

I do admit that the configuration of the bot is complex. I did my best to marriage flexibility and readability, but I think in the future we may eventually redesign this configuration, based on feedback. I'm also open for any suggestions.

sebix commented:
@aaronkaplan you requested some changes to this PR and @kamil-certat implemented various changes based on your feedback. Can you please re-check?

sebix commented:
@aaronkaplan this PR is waiting for your approval

kamil-certat committed:
Regenerate only modified events

aaronkaplan commented:
Still need to test against a misp server

sebix commented:
rebased on develop. there were quite some conflicts to fix.

sebix commented:
now the JSON parser test fails, the output differs to the expected one.

The PR contains changes in `intelmq/lib/message.py`. I don't know yet if this is a rebase error or an error in the code.

sebix commented:
> Still need to test against a misp server

As discussed yesterday, I'll take this over

sebix commented on intelmq/lib/mixins/cache.py:
```diff
@@ -51,6 +67,20 @@ def cache_set(self, key: str, value: Any, ttl: Optional[int] = None):
+    def cache_put(self, value: dict) -> int:
+        # Returns the length of the list after pushing
+        size = self.__redis.lpush(self.bot_id, json.dumps(value))
+        return size
```
```suggestion
    def cache_lpush(self, key: str, value: Any) -> int:
        # Returns the length of the list after pushing
        size = self.__redis.lpush(key, value)
        return size
```

I appreciate the new function, but why is the name of the list not a parameter, but fixed?
Also, why is the value a dict?

The new functions are also missing in the docs in `docs/dev/bot-development.md`

kamil-certat replied:
The idea was to create a message caching for the bot which would make it easy to use, completely abstract from the underlying implementation, not a simple wrapper over the Redis (see e.g. `cache_set`). This is why they do not follow exact the same naming as Redis library, and do additional things.

I have now mixed feelings in which direction it should go. I see two general options:
* renaming my new methods to something more clear, like `cache_dict`, `get_cached_dict`
* go with the simple wrapping over Redis library

I both cases, I would consider also making a key optional (and using bot ID as the default). If going with the first option, then I'd also handle transforming messages to the JSON and back. 

@sebix what do you think is better?

sebix replied:
> * renaming my new methods to something more clear, like `cache_dict`, `get_cached_dict`

These names do not indicate in any way that a FIFO queue is used, but instead make the impression that a dictionary is the cache.
What about `queue_push` and `queue_pop`?

>     * go with the simple wrapping over Redis library

It would be more generic and slightly lower-level. (However, redis/valkey are not that low-level anyway :) )

> I both cases, I would consider also making a key optional (and using bot ID as the default). If going with the first option, then I'd also handle transforming messages to the JSON and back.

I understand your desire for simplicity, but I have doubts about ambiguity. With an optional key, some possible interfaces are:
- `func(key: Optional[str], value: Any)`: Requires calling it with `None` as the first parameter: `func(None, foobar)`. No big difference to using `self.bot_id`: `func(self.bot_id, foobar)`
- `func(value: Any, key: Optional[str] = None)`: The order of parameters is the complete opposite of anything else in IntelMQ and elsewhere.
- `func(key_or_value: Any, value_if_key: Optional[Any])` allowing to call `func(value)` and `func(key, value)`: Too complicated. Not very KISSy :kiss:

kamil-certat replied:
> I understand your desire for simplicity, but I have doubts about ambiguity. With an optional key, some possible interfaces are:

I'd never ever thought about option 3, and the option 1 is not simplifying anything :sweat_smile: So the 2 is the only one I would ever consider

But as my original idea seems to introduce too many questions, I'll move forward with your original idea of wrapping Redis library and let the developer use it flexible.

sebix commented on intelmq/lib/mixins/cache.py:
```diff
@@ -51,6 +67,20 @@ def cache_set(self, key: str, value: Any, ttl: Optional[int] = None):
+        return size
+
+    def cache_length(self) -> int:
+        return self.__redis.llen(self.bot_id)
```
```suggestion
    def cache_llen(self, key: str) -> int:
        return self.__redis.llen(key)
```

The new function is missing in the docs in `docs/dev/bot-development.md`.

sebix commented on intelmq/lib/mixins/cache.py:
```diff
@@ -51,6 +67,20 @@ def cache_set(self, key: str, value: Any, ttl: Optional[int] = None):
+        data = self.__redis.rpop(self.bot_id)
+        if data is None:
+            return None
+        return json.loads(data)
```
```suggestion
    def cache_rpop(self, key: str) -> dict:
        data = self.__redis.rpop(key)
        return data
```

The new function is missing in the docs in `docs/dev/bot-development.md`.

sebix commented on docs/user/bots.md:
```diff
@@ -4760,6 +4766,103 @@ The PyMISP library >= 2.4.119.1 is required, see
+events instead of saving all incoming messages to a one. Each unique value from the field will
+use its own MISP Event. This is useful if your feed provides data about multiple entities you would
+like to group, for example IPs of C2 servers from different botnets. For a given value, the bot will
+use the same MISP Event as long as it's allowed by the `interval_event`.
```
I think this parameter name is misleading and unintuitive. The name "separator" is most often used to specify the element that separates joined data.

From the description, I gather it is actually a grouping or aggregation key.

kamil-certat replied:
`grouping_key` sounds much better, I'll change it

sebix commented on docs/user/bots.md:
````diff
@@ -4760,6 +4766,103 @@ The PyMISP library >= 2.4.119.1 is required, see
+
+For example:
+
+```yaml
+attribute_mapping:
+  source.ip:
+  feed.name:
+    comment: event_description.text
+  destination.ip:
+    to_ids: False
+```
````
An empty value causes an exception:
```
  File "intelmq/bots/outputs/misp/output_feed.py", line 262, in _extract_misp_attribute_kwargs
    for parameter, value in definition.items():
                            ^^^^^^^^^^^^^^^^
AttributeError: 'NoneType' object has no attribute 'items'
```

```suggestion
```yaml
attribute_mapping:
  source.ip: {}
  feed.name:
    comment: event_description.text
  destination.ip:
    to_ids: False
```

sebix reviewed (changes requested):
Found problems in the docs and in the changes of the cache mixin

kamil-certat commented:
I'm working on a few improvements / additional requirements right now, a new code should come today or tomorrow. I'll then address comments

kamil-certat committed:
Add option for flat structure

kamil-certat committed:
Add changing attribute type

kamil-certat committed:
Fix positional arguments

sebix commented:
Documentation for the new parameter `flat_events` is missing

kamil-certat commented:
I'm aware of that, I'll adjust documentation tomorrow

kamil-certat commented:
I also have to add matching unit tests ;)

kamil-certat commented:
BTW, all methods were documented in bot-development :thinking:

sebix commented:
> BTW, all methods were documented in bot-development 🤔

Oh, sorry. I don't know where I was looking.

kamil-certat committed:
Improved documentation, renamed parameters, lower-level cache API

sebix commented on docs/user/bots.md:
```diff
@@ -4772,6 +4772,7 @@ hour", string.
 incoming messages until the given number of them. Use it if your bot proceeds a high number of messages
 and constant saving to the disk is a problem. Reloading or restarting bot as well as generating
 a new MISP event based on `interval_event` triggers regenerating MISP feed regardless of the cache size.
+To ensure saving on reload without any delay, you should also set `_sighup_delay` parameter.
```
That's not a parameter, but an internal variable

kamil-certat committed:
Update Changelog and fixes in docs

kamil-certat committed:
Spelling & pycodestyle

sebix commented on docs/user/bots.md:
```diff
@@ -4760,6 +4766,146 @@ The PyMISP library >= 2.4.119.1 is required, see
+incoming messages until the given number of them. Use it if your bot proceeds a high number of messages
+and constant saving to the disk is a problem. Reloading or restarting bot as well as generating
+a new MISP event based on `interval_event` triggers regenerating MISP feed regardless of the cache size.
+To ensure saving on reload without any delay, you should also set `_sighup_delay` internal variable.
```
What do you want to say with this last sentence?
`_sighup_delay` is not a parameter. It's a switch that developers can use to indicate bot-specific behavior. It is true by default.

Why does the user need to do anything here?

sebix commented on docs/user/bots.md:
```diff
@@ -4760,6 +4766,146 @@ The PyMISP library >= 2.4.119.1 is required, see
+    __all__:
+      - name: tlp:amber
+  # ensure saving on reload
+  _sighup_delay: false
```
Here, you do the opposite of what is written above, disabling it instead of activating it.

And still, this is not something a user should need to do. If the bot needs to disable this flag (maybe depending on other parameters), do it in `init()`.

kamil-certat replied:
This is a good idea, I'll do it so in `init()`. I used wrong wording in the docs. The thing is: if you keep the `_sighup_delay` enabled, then after receiving the reloading signal, the bot waits for the next incoming message to actually do the reloading. This means, the accumulated (cached) events won't be saved. This can be very unfortunate delay

kamil-certat replied:
Actually, I'm stupid, I have already solved it by setting the right defaults... But I'll do it conditionally if using bulk saving

sebix commented on docs/dev/bot-development.md:
```diff
@@ -197,14 +197,48 @@ The `CacheMixin` provides methods to cache values for bots in a Redis database.
+
+Caching key-value pairs and queue caching are two separated mechanisms. The first is designed
+ for arbitrary values, the second one is focused on temporary storing messages (but can handle other
+ data). You won't see caches from one in the another. For example, if adding a key-value pair using
```
```suggestion
Caching key-value pairs and queue caching are two different mechanisms. The functions in the first list are designed
 for arbitrary values, while the latter ones are primarily for temporarily storing messages, but can also handle other data
 types. You won't see caches from one in the other. For example, if adding a key-value pair using
```

sebix commented on docs/dev/bot-development.md:
````diff
@@ -197,14 +197,48 @@ The `CacheMixin` provides methods to cache values for bots in a Redis database.
+
+```python
+msmessage = self.receive_message().to_dict(jsondict_as_string=True)
+self.cache_lpush(self.bot_id, json.dumps(message))
````
Why not use `message.to_json` directly?
```suggestion
self.cache_lpush(self.bot_id, self.receive_message().to_json(jsondict_as_string=True))
```

sebix replied:
Or: `Message.serialize`

kamil-certat replied:
Using the first. I have copied the code from the bot, where the JSON representation is used later.

sebix commented on docs/dev/bot-development.md:
```diff
@@ -197,14 +197,48 @@ The `CacheMixin` provides methods to cache values for bots in a Redis database.
+data = self.cache_pop()
+if data is None:
+    return # handle empty cache
+message = json.loads(data)
+# to use it as Message object
+message_obj = MessageFactory.from_dict(
+    message, harmonization=self.harmonization, default_type="Event"
+)
```
Untested:

```suggestion
message_obj = MessageFactory.unserialize(
    data, harmonization=self.harmonization, default_type="Event"
)
```
I think the two optional parameters are unnecessary too.

kamil-certat replied:
I'm almost sure the default_type was necessary, but I'll check it again.

kamil-certat replied:
The first is recommended, the second is necessary to match the dumping code.

1) if the harmonization is not provided, the code will _read it from the file_, and not use any already loaded version. Reading files and processing JSON is in Python quite expensive operation, it's much better to provide the ready harmonization.
2) the second depends on setting `with_type` when serializing the message. As I didn't set it on serialization, I have to provide type on deserialization. I could do it the other way around, there is no difference - but one or another has to be set.

sebix replied:
Ad 1: True, we should probably cache that file -> different topic
Ad 2: `MessageFactory` is the preferred way of convering Messages to JSON back and forth. I agree, that both ways need to be compatible with each other. Let's refactor this doc section later when improving the overall docs.

kamil-certat committed:
Correct and extend test, automatically adjust reload delay, docs fixes
